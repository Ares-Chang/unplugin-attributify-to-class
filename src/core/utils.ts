export const validateFilterRE = /(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-_:%-?]/

export function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}
