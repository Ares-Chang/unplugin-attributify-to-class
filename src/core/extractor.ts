import type { Options } from '../types'
import { isValidSelector } from './utils'

const strippedPrefixes = [
  'v-bind:',
  ':',
]

const splitterRE = /[\s'"`;]+/g
const elementRE = /<\w(?=.*>)[\w:\.$-]*\s((?:['"`].*?['"`]|.*?)*?)>/gs
const valuedAttributeRE = /([?]|(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-_:!%-]+)(?:=(["'])([^\2]*?)\2)?/g
export const defaultAttributes = ['bg', 'flex', 'grid', 'border', 'text', 'font', 'class', 'className']

interface TransformOption {
  /**
   * 元素标签
   */
  elementStr: string

  /**
   * 元素标签备份
   */
  tempStr: string

  /**
   * 是否有静态class
   */
  staticClass: string

  /**
   *  Attributify转class选择器
   */
  selectors: string[]
}

export const extractorAttributify = (options?: Options): any => {
  const attributes = options?.attributes ?? defaultAttributes
  const nonValuedAttribute = options?.nonValuedAttribute ?? false
  const ignoreNonValuedAttributes = options?.ignoreNonValuedAttributes ?? []
  const prefix = options?.prefix ?? 'un-'
  const prefixedOnly = options?.prefixedOnly ?? false

  return function extract(code: string) {
    const result: TransformOption[] = []

    Array.from(code.matchAll(elementRE))
      .forEach(([elementStr, valuedAttributeStr]) => {
        const valuedAttributes = Array.from((valuedAttributeStr || '').matchAll(valuedAttributeRE))

        const option: TransformOption = {
          elementStr,
          tempStr: elementStr,
          staticClass: '',
          selectors: [],
        }
        valuedAttributes.forEach(([sourceStr, name, _, content]) => {
          const _name = prefixedOnly ? name.replace(prefix, '') : name

          if (!content) {
            // 处理 transform pt2 rounded-sm 单值属性
            if (isValidSelector(name) && nonValuedAttribute !== false) {
              // 不是忽略的非值属性
              if (!ignoreNonValuedAttributes.includes(name)) {
                option.tempStr = option.tempStr.replace(name, '')
                option.selectors.push(name)
              }
            }
            return
          }

          for (const prefix of strippedPrefixes) {
            // 如果是 : v-bind: 开头的属性，则不处理
            if (name.startsWith(prefix)) {
              name = name.slice(prefix.length)
              return
            }
          }

          // 是否生成该属性
          if (!attributes.includes(_name))
            return

          if (['class', 'className'].includes(name)) {
            option.staticClass = sourceStr
          }
          else {
            if (prefixedOnly && !name.startsWith(prefix))
              return

            // 处理 bg="blue-400 hover:blue-500 dark:!blue-500 dark:hover:blue-600"
            const attributifyToClass = content
              .split(splitterRE)
              .filter(Boolean)
              .map(v => `${_name}-${v}`).join(' ')

            option.tempStr = option.tempStr.replace(sourceStr, '')
            option.selectors.push(attributifyToClass)
          }
        })

        result.push(option)
      })

    result.forEach(({ elementStr, tempStr, staticClass, selectors }) => {
      if (selectors.length === 0)
        return

      if (staticClass) {
        tempStr = tempStr.replace(staticClass, spliceStr(staticClass, -1, ` ${selectors.join(' ')}`))
        code = code.replace(elementStr, tempStr)
      }
      else {
        const className = `class="${selectors.join(' ')}"`
        code = code.replace(elementStr, spliceStr(tempStr, -1, className))
      }
    })

    return code
  }
}

function spliceStr(str: string, start: number, newStr: string) {
  return str.slice(0, start) + newStr + str.slice(start)
}
