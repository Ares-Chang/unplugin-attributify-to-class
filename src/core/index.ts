import { defaultRules, transformSelector } from 'unplugin-transform-we-class/utils'
import type { Options } from '../types'

const strippedPrefixes = [
  'v-bind:',
  ':',
  '@',
  'v-on',
]

const templateRe = /<template>([\s\S]*)<\/template>/g
const splitterRE = /[\s'"`;]+/g
const elementRE = /<\w(?=.*>)[\w:\.$-]*\s((?:['"`].*?['"`]|.*?)*?)>/gs
const valuedAttributeRE = /([?]|(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-_:@.!%-]+)(?:=(["'])([^\2]*?)\2)?/g

const validateFilterRE = /(?!\d|-{2}|-\d)[a-zA-Z0-9\u00A0-\uFFFF-_:%-?]/

function isValidSelector(selector = ''): selector is string {
  return validateFilterRE.test(selector)
}

export const defaultAttributes = ['bg', 'flex', 'grid', 'border', 'text', 'font', 'class', 'className', 'p', 'm']
export const defaultIgnoreNonValuedAttributes = ['class']

interface TransformOption {
  /**
   * 元素标签
   */
  elementStr: string

  /**
   * 元素标签备份
   */
  tempStr: string

  /**
   * 是否有静态class
   */
  staticClass: string

  /**
   *  Attributify转class选择器
   */
  selectors: string[]
}

export const extractorAttributify = (options?: Options): any => {
  const attributes = options?.attributes ?? defaultAttributes
  const nonValuedAttribute = options?.nonValuedAttribute ?? false
  const ignoreNonValuedAttributes = options?.ignoreNonValuedAttributes ?? defaultIgnoreNonValuedAttributes
  const prefix = options?.prefix ?? 'un-'
  const prefixedOnly = options?.prefixedOnly ?? false
  const transformEscape = options?.transformEscape ?? true
  const transformRules = options?.transformRules ?? defaultRules

  return function extract(code: string) {
    const result: TransformOption[] = []

    // 只匹配 template 里的代码
    const templateMatchs = Array.from(code.matchAll(templateRe))

    if (templateMatchs.length) {
      const templateCode = templateMatchs[0][1]

      Array.from(templateCode.matchAll(elementRE))
        .forEach(([elementStr, valuedAttributeStr]) => {
          const valuedAttributes = Array.from((valuedAttributeStr || '').matchAll(valuedAttributeRE))

          const option: TransformOption = {
            elementStr,
            tempStr: elementStr,
            staticClass: '',
            selectors: [],
          }
          valuedAttributes.forEach(([sourceStr, name, _, content]) => {
            const _name = prefixedOnly ? name.replace(prefix, '') : name

            if (!content) {
              // 处理 class="" ，生成 2 个 class 属性的问题
              if (name === 'class')
                option.staticClass = sourceStr

              // 处理 transform pt2 rounded-sm 单值属性
              if (isValidSelector(name) && nonValuedAttribute !== false) {
              // 不是忽略的非值属性
                if (!ignoreNonValuedAttributes.includes(name)) {
                // option.tempStr = option.tempStr.replace(name, '')
                  option.selectors.push(transformEscape ? transformSelector(name, transformRules) : name)
                }
              }
              return
            }

            for (const prefix of strippedPrefixes) {
            // 如果是 : v-bind @ v-on 开头的属性，则不处理
              if (name.startsWith(prefix)) {
                name = name.slice(prefix.length)
                return
              }
            }

            // 是否生成该属性
            if (!attributes.includes(_name))
              return

            if (['class', 'className'].includes(name)) {
              option.staticClass = sourceStr
            }

            else {
              if (prefixedOnly && !name.startsWith(prefix))
                return

              // 处理 bg="blue-400"
              const attributifyToClass = content
                .split(splitterRE)
                .filter(Boolean)
                .map(v => v === '~' ? _name : `${_name}-${transformEscape ? transformSelector(v, transformRules) : v}`).join(' ')

              // option.tempStr = option.tempStr.replace(sourceStr, '')
              option.selectors.push(attributifyToClass)
            }
          })

          result.push(option)
        })

      result.forEach(({ elementStr, tempStr, staticClass, selectors }) => {
        if (selectors.length === 0)
          return

        if (staticClass) {
          tempStr = tempStr.replace(staticClass, spliceStr(staticClass, -1, ` ${selectors.join(' ')}`))
          code = code.replace(elementStr, tempStr)
        }
        else {
          const className = ` class="${selectors.join(' ')}"`
          code = code.replace(elementStr, spliceStr(tempStr, -1, className))
        }
      })
    }

    return code
  }
}

function spliceStr(str: string, start: number, newStr: string) {
  return str.slice(0, start) + newStr + str.slice(start)
}
