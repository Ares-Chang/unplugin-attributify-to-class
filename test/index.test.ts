import { describe, expect, test } from 'vitest'
import { extractorAttributify } from '../src/core/extractor'

describe('attributify', () => {
  const fixture1 = `
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg="blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400"
      li-bg="blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500"
      class="text-red font-bold"
      :loading="loading"
      my-class="text-red font-bold"
      title="this is title"
      :class="{ 'text-blue': true, 'text-green': false }"
      :class="[ 'text-blue', 'text-green' ]"
      :class="[ str === 'blue' ? 'text-blue' : 'text-red' ]"
    >
      Button
    </button>

    <image mode="widthFix"></image> 
  </view>
`

  const fixture2 = `
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type="checkbox" peer mt-a>
    <div mb-a group peer-checked="text-4xl">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient="to-r from-yellow-400 via-red-500 to-pink-500"
      >
        ~
      </div>
      <div text-5xl font-100 sm="bg-blue-600">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover="text-teal-400 op-50">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type="text" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder="unocss" 
      un-placeholder="text-red">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top="1/3" transition="200 linear"
        peer-not-placeholder-shown="-translate-y-4 scale-75 origin-top-left text-green-500"
        peer-focus="-translate-y-4 scale-75 origin-top-left text-green-500"
        before="content-!"
        after="content-[!]"
      >Experience now</label>
    </div>
  </div>
</template>
`

  const extract1 = extractorAttributify()
  const extract2 = extractorAttributify({
    prefixedOnly: true,
    prefix: 'li-',
  })

  test('extract1', async () => {
    expect(extract1(fixture1)).toMatchSnapshot()
    expect(extract1(fixture2)).toMatchSnapshot()
  })

  test('extract2', async () => {
    expect(extract2(fixture1)).toMatchSnapshot()
    expect(extract2(fixture2)).toMatchSnapshot()
  })
})
