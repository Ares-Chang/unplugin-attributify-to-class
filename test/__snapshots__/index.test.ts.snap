// Vitest Snapshot v1

exports[`attributify > extract1 1`] = `
"
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      class=\\"text-red font-bold bg-blue-400 bg-hover:blue-400 bg-dark:!blue-400 bg-dark:hover:blue-400\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      
      
      
      
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4 my-attr-y-1 my-attr-x-2 my-attr-sm\\">
      Button
    </button>

    <button 
      
      
      
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button  class=\\"border border-red\\">
      Button
    </button>

    <button  class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a  un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      
      
      
      
      
     class=\\"bg-blue-400 bg-hover:blue-500 bg-dark:blue-500 bg-dark:hover:blue-600 text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      
     class=\\"bg-#333 bg-grey-200 bg-[#444]/40 bg-red/50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
"
`;

exports[`attributify > extract1 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type=\\"checkbox\\" peer mt-a>
    <div mb-a group peer-checked=\\"text-4xl\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
      >
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
      >Experience now</label>
    </div>
  </div>
</template>

<script  lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style></style>
"
`;

exports[`attributify > extract2 1`] = `
"
  <view >
    <button
      h-80 text-center flex flex-col align-center select-none all:transition-400
      bg=\\"blue-400 hover:blue-400 dark:!blue-400 dark:hover:blue-400\\"
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      class=\\"text-red font-bold\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      non-value
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      my-attr=\\"y-1 x-2 sm\\"
    >
      Button
    </button>

    <button 
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
    >
      Button
    </button>

    <button border=\\"~ red\\">
      Button
    </button>

    <button flex=\\"~ col wrap\\">
      Button
    </button>

    <a text=\\"red\\"  class=\\"text-blue\\">This conflicts with links' text prop</a>

    <button 
      bg=\\"blue-400 hover:blue-500 dark:blue-500 dark:hover:blue-600\\"
      text=\\"sm white\\"
      font=\\"mono light\\"
      p=\\"y-2 x-4\\"
      border=\\"2 rounded blue-200\\"
    >
      Button
    </button>

    <button 
      bg=\\"#333 grey-200 [#444]/40 red/50\\"
    >
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
"
`;

exports[`attributify > extract2 2`] = `
"
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type=\\"checkbox\\" peer mt-a>
    <div mb-a group peer-checked=\\"text-4xl\\">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
      >
        ~
      </div>
      <div text-5xl font-100 sm=\\"bg-blue-600\\">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover=\\"text-teal-400 op-50\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type=\\"text\\" m-0 pt-4 px-4 text-true-gray-800 peer 
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\">
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
      >Experience now</label>
    </div>
  </div>
</template>

<script  lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style></style>
"
`;

exports[`attributify > extract3 1`] = `
"
  <view >
    <button
      h-80  flex    
      
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      class=\\"text-red font-bold text-center flex-col align-center select-none all:transition-400 bg-blue-400 bg-hover:blue-400 bg-dark:!blue-400 bg-dark:hover:blue-400 non-value\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      my-prop
      
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      
      
      
      my-attr=\\"y-1 x-2 sm\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button 
      
      
      
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button  class=\\"border border-red\\">
      Button
    </button>

    <button  class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a  un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      
      
      
      
      
     class=\\"bg-blue-400 bg-hover:blue-500 bg-dark:blue-500 bg-dark:hover:blue-600 text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      
     class=\\"bg-#333 bg-grey-200 bg-[#444]/40 bg-red/50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
"
`;

exports[`attributify > extract3 2`] = `
"
<template>
  <div h-80  flex     class=\\"text-center flex-col align-center select-none all:transition-400\\">
    <input type=\\"checkbox\\"   class=\\"peer mt-a\\">
    <div   peer-checked=\\"text-4xl\\" class=\\"mb-a group\\">
      <div
           
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
       class=\\"font-100 text-4xl mb--3 p-10\\">
        ~
      </div>
      <div   sm=\\"bg-blue-600\\" class=\\"text-5xl font-100\\">
        unocss
      </div>
      <div     group-hover=\\"text-teal-400 op-50\\" class=\\"op-20 font-200 mt-1 tracking-wider\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div     class=\\"ma inline-flex relative h-14\\">
      <input type=\\"text\\"      
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\" class=\\"m-0 pt-4 px-4 text-true-gray-800 peer\\">
      <label      top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
       class=\\"absolute leading-1rem left-4 pointer-events-none text-gray-7\\">Experience now</label>
    </div>
  </div>
</template>

<script  lang=\\"ts\\"  class=\\"setup\\">
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style></style>
"
`;

exports[`attributify > extract4 1`] = `
"
  <view >
    <button
            
      
      li-bg=\\"blue-500 hover:blue-500 dark:!blue-500 dark:hover:blue-500\\"
      class=\\"text-red font-bold h-80 text-center flex flex-col align-center select-none all:transition-400 bg-blue-400 bg-hover:blue-400 bg-dark:!blue-400 bg-dark:hover:blue-400 my-prop non-value\\"
      :loading=\\"loading\\"
      my-class=\\"text-red font-bold\\"
      title=\\"this is title\\"
      
      
      :class=\\"{ 'text-blue': true, 'text-green': false }\\"
      :class=\\"[ 'text-blue', 'text-green' ]\\"
      :class=\\"[ str === 'blue' ? 'text-blue' : 'text-red' ]\\"
    >
      Button
    </button>

    <button 
      
      
      
      my-attr=\\"y-1 x-2 sm\\"
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button 
      
      
      
     class=\\"text-sm text-white font-mono font-light p-y-2 p-x-4\\">
      Button
    </button>

    <button  class=\\"border border-red\\">
      Button
    </button>

    <button  class=\\"flex flex-col flex-wrap\\">
      Button
    </button>

    <a  un-text=\\"blue\\" class=\\"text-red\\">This conflicts with links' text prop</a>

    <button 
      
      
      
      
      
     class=\\"bg-blue-400 bg-hover:blue-500 bg-dark:blue-500 bg-dark:hover:blue-600 text-sm text-white font-mono font-light p-y-2 p-x-4 border-2 border-rounded border-blue-200\\">
      Button
    </button>

    <button 
      
     class=\\"bg-#333 bg-grey-200 bg-[#444]/40 bg-red/50\\">
      Button
    </button>

    <image mode=\\"widthFix\\"></image> 
  </view>
"
`;

exports[`attributify > extract4 2`] = `
"
<template>
  <div        class=\\"h-80 text-center flex flex-col align-center select-none all:transition-400\\">
    <input type=\\"checkbox\\"   class=\\"peer mt-a\\">
    <div   peer-checked=\\"text-4xl\\" class=\\"mb-a group\\">
      <div
           
        bg-gradient=\\"to-r from-yellow-400 via-red-500 to-pink-500\\"
       class=\\"font-100 text-4xl mb--3 p-10\\">
        ~
      </div>
      <div   sm=\\"bg-blue-600\\" class=\\"text-5xl font-100\\">
        unocss
      </div>
      <div     group-hover=\\"text-teal-400 op-50\\" class=\\"op-20 font-200 mt-1 tracking-wider\\">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div  class=\\"flex\\">
    <div     class=\\"ma inline-flex relative h-14\\">
      <input type=\\"text\\"      
      placeholder=\\"unocss\\" 
      un-placeholder=\\"text-red\\" class=\\"m-0 pt-4 px-4 text-true-gray-800 peer\\">
      <label      top=\\"1/3\\" transition=\\"200 linear\\"
        peer-not-placeholder-shown=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        peer-focus=\\"-translate-y-4 scale-75 origin-top-left text-green-500\\"
        before=\\"content-!\\"
        after=\\"content-[!]\\"
       class=\\"absolute leading-1rem left-4 pointer-events-none text-gray-7\\">Experience now</label>
    </div>
  </div>
</template>

<script  lang=\\"ts\\" setup>
import { onHide, onLaunch, onShow } from '@dcloudio/uni-app'
onLaunch(() => {
  console.log('App Launch')
})
onShow(() => {
  console.log('App Show')
})
onHide(() => {
  console.log('App Hide')
})
</script>

<style></style>
"
`;
